need `__from_arrow__ ` to support `types_mapper`

I attempting to convert from Arrow to Pandas using the `types_mapper` argument of [`to_pandas`](https://arrow.apache.org/docs/python/generated/pyarrow.RecordBatch.html#pyarrow.RecordBatch.to_pandas), I get the following error:

```
________________ test_list_rows_nullable_scalars_dtypes[None] ___________________

bigquery_client = <google.cloud.bigquery.client.Client object at 0x197b50a30>
scalars_table = 'swast-scratch.python_bigquery_tests_system_20211102220533_55dcac.scalars', max_results = None

    @pytest.mark.parametrize(
        ("max_results",), ((None,), (10,),)  # Use BQ Storage API.  # Use REST API.
    )
    def test_list_rows_nullable_scalars_dtypes(bigquery_client, scalars_table, max_results):
        # TODO(GH#836): Avoid INTERVAL columns until they are supported by the
        # BigQuery Storage API and pyarrow.
        schema = [
            bigquery.SchemaField("bool_col", enums.SqlTypeNames.BOOLEAN),
            bigquery.SchemaField("bignumeric_col", enums.SqlTypeNames.BIGNUMERIC),
            bigquery.SchemaField("bytes_col", enums.SqlTypeNames.BYTES),
            bigquery.SchemaField("date_col", enums.SqlTypeNames.DATE),
            bigquery.SchemaField("datetime_col", enums.SqlTypeNames.DATETIME),
            bigquery.SchemaField("float64_col", enums.SqlTypeNames.FLOAT64),
            bigquery.SchemaField("geography_col", enums.SqlTypeNames.GEOGRAPHY),
            bigquery.SchemaField("int64_col", enums.SqlTypeNames.INT64),
            bigquery.SchemaField("numeric_col", enums.SqlTypeNames.NUMERIC),
            bigquery.SchemaField("string_col", enums.SqlTypeNames.STRING),
            bigquery.SchemaField("time_col", enums.SqlTypeNames.TIME),
            bigquery.SchemaField("timestamp_col", enums.SqlTypeNames.TIMESTAMP),
        ]
    
>       df = bigquery_client.list_rows(
            scalars_table, max_results=max_results, selected_fields=schema,
        ).to_dataframe()

tests/system/test_pandas.py:1030: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
google/cloud/bigquery/table.py:1946: in to_dataframe
    df = record_batch.to_pandas(
pyarrow/array.pxi:757: in pyarrow.lib._PandasConvertible.to_pandas
    ???
pyarrow/table.pxi:1740: in pyarrow.lib.Table._to_pandas
    ???
/usr/local/Caskroom/miniconda/base/envs/dev-3.9/lib/python3.9/site-packages/pyarrow/pandas_compat.py:789: in table_to_blockmanager
    blocks = _table_to_blocks(options, table, categories, ext_columns_dtypes)
/usr/local/Caskroom/miniconda/base/envs/dev-3.9/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1130: in _table_to_blocks
    return [_reconstruct_block(item, columns, extension_columns)
/usr/local/Caskroom/miniconda/base/envs/dev-3.9/lib/python3.9/site-packages/pyarrow/pandas_compat.py:1130: in <listcomp>
    return [_reconstruct_block(item, columns, extension_columns)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

item = {'placement': array([3]), 'py_array': <pyarrow.lib.ChunkedArray object at 0x197c53db0>
[
  [
    2021-07-21,
    null
  ]
]}
columns = ['bool_col', 'bignumeric_col', 'bytes_col', 'date_col', 'datetime_col', 'float64_col', ...]
extension_columns = {'date_col': <class 'db_dtypes.DateDtype'>}

    def _reconstruct_block(item, columns=None, extension_columns=None):
        """
        Construct a pandas Block from the `item` dictionary coming from pyarrow's
        serialization or returned by arrow::python::ConvertTableToPandas.
    
        This function takes care of converting dictionary types to pandas
        categorical, Timestamp-with-timezones to the proper pandas Block, and
        conversion to pandas ExtensionBlock
    
        Parameters
        ----------
        item : dict
            For basic types, this is a dictionary in the form of
            {'block': np.ndarray of values, 'placement': pandas block placement}.
            Additional keys are present for other types (dictionary, timezone,
            object).
        columns :
            Column names of the table being constructed, used for extension types
        extension_columns : dict
            Dictionary of {column_name: pandas_dtype} that includes all columns
            and corresponding dtypes that will be converted to a pandas
            ExtensionBlock.
    
        Returns
        -------
        pandas Block
    
        """
        import pandas.core.internals as _int
    
        block_arr = item.get('block', None)
        placement = item['placement']
        if 'dictionary' in item:
            cat = _pandas_api.categorical_type.from_codes(
                block_arr, categories=item['dictionary'],
                ordered=item['ordered'])
            block = _int.make_block(cat, placement=placement)
        elif 'timezone' in item:
            dtype = make_datetimetz(item['timezone'])
            block = _int.make_block(block_arr, placement=placement,
                                    klass=_int.DatetimeTZBlock,
                                    dtype=dtype)
        elif 'object' in item:
            block = _int.make_block(builtin_pickle.loads(block_arr),
                                    placement=placement)
        elif 'py_array' in item:
            # create ExtensionBlock
            arr = item['py_array']
            assert len(placement) == 1
            name = columns[placement[0]]
            pandas_dtype = extension_columns[name]
            if not hasattr(pandas_dtype, '__from_arrow__'):
>               raise ValueError("This column does not support to be converted "
                                 "to a pandas ExtensionArray")
E               ValueError: This column does not support to be converted to a pandas ExtensionArray

/usr/local/Caskroom/miniconda/base/envs/dev-3.9/lib/python3.9/site-packages/pyarrow/pandas_compat.py:747: ValueError
================== short test summary info ================
```

Context: https://github.com/googleapis/python-bigquery/pull/972/files#r741469631

--- a/db_dtypes/__init__.py
+++ b/db_dtypes/__init__.py
@@ -17,6 +17,7 @@ Pandas Data Types for SQL systems (BigQuery, Spanner)
 
 import datetime
 import re
+from typing import Union
 
 import numpy
 import packaging.version
@@ -52,6 +53,13 @@ class TimeDtype(core.BaseDatetimeDtype):
     def construct_array_type(self):
         return TimeArray
 
+    @staticmethod
+    def __from_arrow__(
+        array: Union[pyarrow.Array, pyarrow.ChunkedArray]
+    ) -> "TimeArray":
+        # TODO: Consider a more efficient conversion to "M8[ns]" numpy array.
+        return TimeArray(array)
+
 
 class TimeArray(core.BaseDatetimeArray):
     """
@@ -75,7 +83,13 @@ class TimeArray(core.BaseDatetimeArray):
             r"(?:\.(?P<fraction>\d*))?)?)?\s*$"
         ).match,
     ):
-        if isinstance(scalar, datetime.time):
+        # Convert pyarrow values to datetime.time.
+        if isinstance(scalar, (pyarrow.Time32Scalar, pyarrow.Time64Scalar)):
+            scalar = scalar.as_py()
+
+        if scalar is None:
+            return None
+        elif isinstance(scalar, datetime.time):
             return datetime.datetime.combine(cls._epoch, scalar)
         elif isinstance(scalar, str):
             # iso string
@@ -146,6 +160,13 @@ class DateDtype(core.BaseDatetimeDtype):
     def construct_array_type(self):
         return DateArray
 
+    @staticmethod
+    def __from_arrow__(
+        array: Union[pyarrow.Array, pyarrow.ChunkedArray]
+    ) -> "DateArray":
+        # TODO: Consider a more efficient conversion to "M8[ns]" numpy array.
+        return DateArray(array)
+
 
 class DateArray(core.BaseDatetimeArray):
     """
@@ -161,7 +182,13 @@ class DateArray(core.BaseDatetimeArray):
         scalar,
         match_fn=re.compile(r"\s*(?P<year>\d+)-(?P<month>\d+)-(?P<day>\d+)\s*$").match,
     ):
-        if isinstance(scalar, datetime.date):
+        # Convert pyarrow values to datetime.date.
+        if isinstance(scalar, (pyarrow.Date32Scalar, pyarrow.Date64Scalar)):
+            scalar = scalar.as_py()
+
+        if scalar is None:
+            return None
+        elif isinstance(scalar, datetime.date):
             return datetime.datetime(scalar.year, scalar.month, scalar.day)
         elif isinstance(scalar, str):
             match = match_fn(scalar)

--- a/db_dtypes/__init__.py
+++ b/db_dtypes/__init__.py
@@ -60,6 +60,11 @@ class TimeDtype(core.BaseDatetimeDtype):
     def __from_arrow__(
         array: Union[pyarrow.Array, pyarrow.ChunkedArray]
     ) -> "TimeArray":
+        """Convert to dbtime data from an Arrow array.
+
+        See:
+        https://pandas.pydata.org/pandas-docs/stable/development/extending.html#compatibility-with-apache-arrow
+        """
         # We can't call combine_chunks on an empty array, so short-circuit the
         # rest of the function logic for this special case.
         if len(array) == 0:
@@ -153,6 +158,11 @@ class TimeArray(core.BaseDatetimeArray):
             return super().astype(dtype, copy=copy)
 
     def __arrow_array__(self, type=None):
+        """Convert to an Arrow array from dbtime data.
+
+        See:
+        https://pandas.pydata.org/pandas-docs/stable/development/extending.html#compatibility-with-apache-arrow
+        """
         array = pyarrow.array(self._ndarray, type=pyarrow.timestamp("ns"))
 
         # ChunkedArray has no "view" method, so combine into an Array.
@@ -185,6 +195,11 @@ class DateDtype(core.BaseDatetimeDtype):
     def __from_arrow__(
         array: Union[pyarrow.Array, pyarrow.ChunkedArray]
     ) -> "DateArray":
+        """Convert to dbdate data from an Arrow array.
+
+        See:
+        https://pandas.pydata.org/pandas-docs/stable/development/extending.html#compatibility-with-apache-arrow
+        """
         array = pyarrow.compute.cast(array, pyarrow.timestamp("ns"))
         np_array = array.to_numpy()
         return DateArray(np_array)
@@ -246,6 +261,11 @@ class DateArray(core.BaseDatetimeArray):
         return super().astype(dtype, copy=copy)
 
     def __arrow_array__(self, type=None):
+        """Convert to an Arrow array from dbdate data.
+
+        See:
+        https://pandas.pydata.org/pandas-docs/stable/development/extending.html#compatibility-with-apache-arrow
+        """
         array = pyarrow.array(self._ndarray, type=pyarrow.timestamp("ns"))
         return pyarrow.compute.cast(
             array, type if type is not None else pyarrow.date32(),

--- a/db_dtypes/__init__.py
+++ b/db_dtypes/__init__.py
@@ -40,7 +40,6 @@ date_dtype_name = "dbdate"
 time_dtype_name = "dbtime"
 _EPOCH = datetime.datetime(1970, 1, 1)
 _NPEPOCH = numpy.datetime64(_EPOCH)
-_PAEPOCH = pyarrow.scalar(_EPOCH, pyarrow.timestamp("ns"))
 
 pandas_release = packaging.version.parse(pandas.__version__).release
 
@@ -66,7 +65,7 @@ class TimeDtype(core.BaseDatetimeDtype):
         if len(array) == 0:
             return TimeArray(numpy.array([], dtype="datetime64[ns]"))
 
-        # We can't cast to timestamp("ns"), but we time64("ns") has the same
+        # We can't cast to timestamp("ns"), but time64("ns") has the same
         # memory layout: 64-bit integers representing the number of nanoseconds
         # since the datetime epoch (midnight 1970-01-01).
         array = pyarrow.compute.cast(array, pyarrow.time64("ns"))
@@ -157,9 +156,17 @@ class TimeArray(core.BaseDatetimeArray):
             return self.astype(dtype)
 
     def __arrow_array__(self, type=None):
-        return pyarrow.array(
-            self.to_numpy(dtype="object"),
-            type=type if type is not None else pyarrow.time64("ns"),
+        array = pyarrow.array(self._ndarray, type=pyarrow.timestamp("ns"))
+
+        # ChunkedArray has no "view" method, so combine into an Array.
+        array = array.combine_chunks() if hasattr(array, "combine_chunks") else array
+
+        # We can't cast to time64("ns"), but timestamp("ns") has the same
+        # memory layout: 64-bit integers representing the number of nanoseconds
+        # since the datetime epoch (midnight 1970-01-01).
+        array = array.view(pyarrow.time64("ns"))
+        return pyarrow.compute.cast(
+            array, type if type is not None else pyarrow.time64("ns"),
         )
 
 
@@ -240,8 +247,9 @@ class DateArray(core.BaseDatetimeArray):
         return super().astype(dtype, copy=copy)
 
     def __arrow_array__(self, type=None):
-        return pyarrow.array(
-            self._ndarray, type=type if type is not None else pyarrow.date32(),
+        array = pyarrow.array(self._ndarray, type=pyarrow.timestamp("ns"))
+        return pyarrow.compute.cast(
+            array, type if type is not None else pyarrow.date32(),
         )
 
     def __add__(self, other):

--- a/db_dtypes/__init__.py
+++ b/db_dtypes/__init__.py
@@ -150,11 +150,6 @@ class TimeArray(core.BaseDatetimeArray):
         else:
             return super().astype(dtype, copy=copy)
 
-    if pandas_release < (1,):
-
-        def to_numpy(self, dtype="object"):
-            return self.astype(dtype)
-
     def __arrow_array__(self, type=None):
         array = pyarrow.array(self._ndarray, type=pyarrow.timestamp("ns"))
 

--- a/db_dtypes/__init__.py
+++ b/db_dtypes/__init__.py
@@ -71,7 +71,9 @@ class TimeDtype(core.BaseDatetimeDtype):
         array = pyarrow.compute.cast(array, pyarrow.time64("ns"))
 
         # ChunkedArray has no "view" method, so combine into an Array.
-        array = array.combine_chunks() if hasattr(array, "combine_chunks") else array
+        if isinstance(array, pyarrow.ChunkedArray):
+            array = array.combine_chunks()
+
         array = array.view(pyarrow.timestamp("ns"))
         np_array = array.to_numpy(zero_copy_only=False)
         return TimeArray(np_array)
@@ -154,7 +156,9 @@ class TimeArray(core.BaseDatetimeArray):
         array = pyarrow.array(self._ndarray, type=pyarrow.timestamp("ns"))
 
         # ChunkedArray has no "view" method, so combine into an Array.
-        array = array.combine_chunks() if hasattr(array, "combine_chunks") else array
+        array = (
+            array.combine_chunks() if isinstance(array, pyarrow.ChunkedArray) else array
+        )
 
         # We can't cast to time64("ns"), but timestamp("ns") has the same
         # memory layout: 64-bit integers representing the number of nanoseconds

--- a/db_dtypes/__init__.py
+++ b/db_dtypes/__init__.py
@@ -30,6 +30,7 @@ import pandas.core.dtypes.dtypes
 import pandas.core.dtypes.generic
 import pandas.core.nanops
 import pyarrow
+import pyarrow.compute
 
 from db_dtypes.version import __version__
 from db_dtypes import core
@@ -37,6 +38,9 @@ from db_dtypes import core
 
 date_dtype_name = "dbdate"
 time_dtype_name = "dbtime"
+_EPOCH = datetime.datetime(1970, 1, 1)
+_NPEPOCH = numpy.datetime64(_EPOCH)
+_PAEPOCH = pyarrow.scalar(_EPOCH, pyarrow.timestamp("ns"))
 
 pandas_release = packaging.version.parse(pandas.__version__).release
 
@@ -57,8 +61,21 @@ class TimeDtype(core.BaseDatetimeDtype):
     def __from_arrow__(
         array: Union[pyarrow.Array, pyarrow.ChunkedArray]
     ) -> "TimeArray":
-        # TODO: Consider a more efficient conversion to "M8[ns]" numpy array.
-        return TimeArray(array)
+        # We can't call combine_chunks on an empty array, so short-circuit the
+        # rest of the function logic for this special case.
+        if len(array) == 0:
+            return TimeArray(numpy.array([], dtype="datetime64[ns]"))
+
+        # We can't cast to timestamp("ns"), but we time64("ns") has the same
+        # memory layout: 64-bit integers representing the number of nanoseconds
+        # since the datetime epoch (midnight 1970-01-01).
+        array = pyarrow.compute.cast(array, pyarrow.time64("ns"))
+
+        # ChunkedArray has no "view" method, so combine into an Array.
+        array = array.combine_chunks() if hasattr(array, "combine_chunks") else array
+        array = array.view(pyarrow.timestamp("ns"))
+        np_array = array.to_numpy(zero_copy_only=False)
+        return TimeArray(np_array)
 
 
 class TimeArray(core.BaseDatetimeArray):
@@ -69,8 +86,6 @@ class TimeArray(core.BaseDatetimeArray):
     # Data are stored as datetime64 values with a date of Jan 1, 1970
 
     dtype = TimeDtype()
-    _epoch = datetime.datetime(1970, 1, 1)
-    _npepoch = numpy.datetime64(_epoch)
 
     @classmethod
     def _datetime(
@@ -90,7 +105,7 @@ class TimeArray(core.BaseDatetimeArray):
         if scalar is None:
             return None
         elif isinstance(scalar, datetime.time):
-            return datetime.datetime.combine(cls._epoch, scalar)
+            return datetime.datetime.combine(_EPOCH, scalar)
         elif isinstance(scalar, str):
             # iso string
             parsed = match_fn(scalar)
@@ -127,7 +142,7 @@ class TimeArray(core.BaseDatetimeArray):
     __return_deltas = {"timedelta", "timedelta64", "timedelta64[ns]", "<m8", "<m8[ns]"}
 
     def astype(self, dtype, copy=True):
-        deltas = self._ndarray - self._npepoch
+        deltas = self._ndarray - _NPEPOCH
         stype = str(dtype)
         if stype in self.__return_deltas:
             return deltas
@@ -164,8 +179,9 @@ class DateDtype(core.BaseDatetimeDtype):
     def __from_arrow__(
         array: Union[pyarrow.Array, pyarrow.ChunkedArray]
     ) -> "DateArray":
-        # TODO: Consider a more efficient conversion to "M8[ns]" numpy array.
-        return DateArray(array)
+        array = pyarrow.compute.cast(array, pyarrow.timestamp("ns"))
+        np_array = array.to_numpy()
+        return DateArray(np_array)
 
 
 class DateArray(core.BaseDatetimeArray):
@@ -233,7 +249,7 @@ class DateArray(core.BaseDatetimeArray):
             return self.astype("object") + other
 
         if isinstance(other, TimeArray):
-            return (other._ndarray - other._npepoch) + self._ndarray
+            return (other._ndarray - _NPEPOCH) + self._ndarray
 
         return super().__add__(other)
 

--- a/db_dtypes/__init__.py
+++ b/db_dtypes/__init__.py
@@ -106,12 +106,19 @@ class TimeArray(core.BaseDatetimeArray):
     ):
         # Convert pyarrow values to datetime.time.
         if isinstance(scalar, (pyarrow.Time32Scalar, pyarrow.Time64Scalar)):
-            scalar = scalar.as_py()
+            scalar = (
+                scalar.cast(pyarrow.time64("ns"))
+                .cast(pyarrow.int64())
+                .cast(pyarrow.timestamp("ns"))
+                .as_py()
+            )
 
         if scalar is None:
             return None
         elif isinstance(scalar, datetime.time):
             return datetime.datetime.combine(_EPOCH, scalar)
+        elif isinstance(scalar, pandas.Timestamp):
+            return scalar.to_datetime64()
         elif isinstance(scalar, str):
             # iso string
             parsed = match_fn(scalar)

--- a/db_dtypes/core.py
+++ b/db_dtypes/core.py
@@ -61,10 +61,7 @@ class BaseDatetimeArray(
 
     @classmethod
     def __ndarray(cls, scalars):
-        return numpy.array(
-            [None if scalar is None else cls._datetime(scalar) for scalar in scalars],
-            "M8[ns]",
-        )
+        return numpy.array([cls._datetime(scalar) for scalar in scalars], "M8[ns]",)
 
     @classmethod
     def _from_sequence(cls, scalars, *, dtype=None, copy=False):

--- a/db_dtypes/core.py
+++ b/db_dtypes/core.py
@@ -17,6 +17,7 @@ from typing import Any, Optional, Sequence
 import numpy
 import pandas
 from pandas._libs import NaT
+import pandas.api.extensions
 import pandas.compat.numpy.function
 import pandas.core.algorithms
 import pandas.core.arrays
@@ -32,7 +33,7 @@ from db_dtypes import pandas_backports
 pandas_release = pandas_backports.pandas_release
 
 
-class BaseDatetimeDtype(pandas.core.dtypes.base.ExtensionDtype):
+class BaseDatetimeDtype(pandas.api.extensions.ExtensionDtype):
     na_value = NaT
     kind = "o"
     names = None

--- a/tests/unit/test_arrow.py
+++ b/tests/unit/test_arrow.py
@@ -24,6 +24,11 @@ import pytest
 import db_dtypes
 
 
+SECOND_NANOS = 1_000_000_000
+MINUTE_NANOS = 60 * SECOND_NANOS
+HOUR_NANOS = 60 * MINUTE_NANOS
+
+
 def types_mapper(
     pyarrow_type: pyarrow.DataType,
 ) -> Optional[pandas.api.extensions.ExtensionDtype]:
@@ -158,6 +163,40 @@ SERIES_ARRAYS_CUSTOM_ARROW_TYPES = [
                 dt.time(23, 59, 59, 999_999),
             ],
             type=pyarrow.time64("us"),
+        ),
+    ),
+    (
+        pandas.Series(
+            [
+                # Only microseconds are supported when reading data. See:
+                # https://github.com/googleapis/python-db-dtypes-pandas/issues/19
+                # Still, round-trip with pyarrow nanosecond precision scalars
+                # is supported.
+                pyarrow.scalar(0, pyarrow.time64("ns")),
+                pyarrow.scalar(
+                    12 * HOUR_NANOS
+                    + 30 * MINUTE_NANOS
+                    + 15 * SECOND_NANOS
+                    + 123_456_789,
+                    pyarrow.time64("ns"),
+                ),
+                pyarrow.scalar(
+                    23 * HOUR_NANOS
+                    + 59 * MINUTE_NANOS
+                    + 59 * SECOND_NANOS
+                    + 999_999_999,
+                    pyarrow.time64("ns"),
+                ),
+            ],
+            dtype="dbtime",
+        ),
+        pyarrow.array(
+            [
+                0,
+                12 * HOUR_NANOS + 30 * MINUTE_NANOS + 15 * SECOND_NANOS + 123_456_789,
+                23 * HOUR_NANOS + 59 * MINUTE_NANOS + 59 * SECOND_NANOS + 999_999_999,
+            ],
+            type=pyarrow.time64("ns"),
         ),
     ),
 ]

--- a/tests/unit/test_arrow.py
+++ b/tests/unit/test_arrow.py
@@ -140,6 +140,24 @@ SERIES_ARRAYS_CUSTOM_ARROW_TYPES = [
     ),
     (
         pandas.Series(
+            [
+                dt.time(0, 0, 0, 0),
+                dt.time(12, 30, 15, 125_000),
+                dt.time(23, 59, 59, 999_000),
+            ],
+            dtype="dbtime",
+        ),
+        pyarrow.array(
+            [
+                dt.time(0, 0, 0, 0),
+                dt.time(12, 30, 15, 125_000),
+                dt.time(23, 59, 59, 999_000),
+            ],
+            type=pyarrow.time32("ms"),
+        ),
+    ),
+    (
+        pandas.Series(
             [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_999)], dtype="dbtime",
         ),
         pyarrow.array(

--- a/tests/unit/test_arrow.py
+++ b/tests/unit/test_arrow.py
@@ -203,6 +203,29 @@ def test_series_from_arrow_scalars(
     pandas.testing.assert_series_equal(series, expected)
 
 
+def test_dbtime_series_from_arrow_array():
+    """Test to explicitly check Array -> Series conversion."""
+    array = pyarrow.array([dt.time(15, 21, 0, 123_456)], type=pyarrow.time64("us"))
+    assert isinstance(array, pyarrow.Array)
+    assert not isinstance(array, pyarrow.ChunkedArray)
+    series = pandas.Series(db_dtypes.TimeDtype.__from_arrow__(array))
+    expected = pandas.Series([dt.time(15, 21, 0, 123_456)], dtype="dbtime")
+    pandas.testing.assert_series_equal(series, expected)
+
+
+def test_dbtime_series_from_arrow_chunkedarray():
+    """Test to explicitly check ChunkedArray -> Series conversion."""
+    array1 = pyarrow.array([dt.time(15, 21, 0, 123_456)], type=pyarrow.time64("us"))
+    array2 = pyarrow.array([dt.time(0, 0, 0, 0)], type=pyarrow.time64("us"))
+    array = pyarrow.chunked_array([array1, array2])
+    assert isinstance(array, pyarrow.ChunkedArray)
+    series = pandas.Series(db_dtypes.TimeDtype.__from_arrow__(array))
+    expected = pandas.Series(
+        [dt.time(15, 21, 0, 123_456), dt.time(0, 0, 0, 0)], dtype="dbtime"
+    )
+    pandas.testing.assert_series_equal(series, expected)
+
+
 def test_dataframe_from_arrow():
     record_batch = pyarrow.RecordBatch.from_arrays(
         [

--- a/tests/unit/test_arrow.py
+++ b/tests/unit/test_arrow.py
@@ -185,4 +185,4 @@ def test_from_arrow(pyarrow_array: pyarrow.Array, expected: pandas.Series):
     record_batch = pyarrow.RecordBatch.from_arrays([pyarrow_array], ["test_col"])
     dataframe = record_batch.to_pandas(date_as_object=False, types_mapper=types_mapper)
     series = dataframe["test_col"]
-    pandas.testing.assert_extension_array_equal(series, expected)
+    pandas.testing.assert_series_equal(series, expected, check_names=False)

--- a/tests/unit/test_arrow.py
+++ b/tests/unit/test_arrow.py
@@ -13,160 +13,176 @@
 # limitations under the License.
 
 import datetime as dt
+from typing import Optional
 
 import pandas
+import pandas.api.extensions
+import pandas.testing
 import pyarrow
 import pytest
 
-# To register the types.
-import db_dtypes  # noqa
+import db_dtypes
 
 
-@pytest.mark.parametrize(
-    ("series", "expected"),
+def types_mapper(
+    pyarrow_type: pyarrow.DataType,
+) -> Optional[pandas.api.extensions.ExtensionDtype]:
+    type_str = str(pyarrow_type)
+
+    if type_str.startswith("date32") or type_str.startswith("date64"):
+        return db_dtypes.DateDtype
+    elif type_str.startswith("time32") or type_str.startswith("time64"):
+        return db_dtypes.TimeDtype
+    else:
+        # Use default type mapping.
+        return None
+
+
+SERIES_ARRAYS_DEFAULT_TYPES = [
+    (pandas.Series([], dtype="dbdate"), pyarrow.array([], type=pyarrow.date32())),
     (
-        (pandas.Series([], dtype="dbdate"), pyarrow.array([], type=pyarrow.date32())),
-        (
-            pandas.Series([None, None, None], dtype="dbdate"),
-            pyarrow.array([None, None, None], type=pyarrow.date32()),
+        pandas.Series([None, None, None], dtype="dbdate"),
+        pyarrow.array([None, None, None], type=pyarrow.date32()),
+    ),
+    (
+        pandas.Series(
+            [dt.date(2021, 9, 27), None, dt.date(2011, 9, 27)], dtype="dbdate"
         ),
-        (
-            pandas.Series(
-                [dt.date(2021, 9, 27), None, dt.date(2011, 9, 27)], dtype="dbdate"
-            ),
-            pyarrow.array(
-                [dt.date(2021, 9, 27), None, dt.date(2011, 9, 27)],
-                type=pyarrow.date32(),
-            ),
-        ),
-        (
-            pandas.Series(
-                [dt.date(1677, 9, 22), dt.date(1970, 1, 1), dt.date(2262, 4, 11)],
-                dtype="dbdate",
-            ),
-            pyarrow.array(
-                [dt.date(1677, 9, 22), dt.date(1970, 1, 1), dt.date(2262, 4, 11)],
-                type=pyarrow.date32(),
-            ),
-        ),
-        (
-            pandas.Series([], dtype="dbtime"),
-            pyarrow.array([], type=pyarrow.time64("ns")),
-        ),
-        (
-            pandas.Series([None, None, None], dtype="dbtime"),
-            pyarrow.array([None, None, None], type=pyarrow.time64("ns")),
-        ),
-        (
-            pandas.Series(
-                [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_999)],
-                dtype="dbtime",
-            ),
-            pyarrow.array(
-                [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_999)],
-                type=pyarrow.time64("ns"),
-            ),
-        ),
-        (
-            pandas.Series(
-                [
-                    dt.time(0, 0, 0, 0),
-                    dt.time(12, 30, 15, 125_000),
-                    dt.time(23, 59, 59, 999_999),
-                ],
-                dtype="dbtime",
-            ),
-            pyarrow.array(
-                [
-                    dt.time(0, 0, 0, 0),
-                    dt.time(12, 30, 15, 125_000),
-                    dt.time(23, 59, 59, 999_999),
-                ],
-                type=pyarrow.time64("ns"),
-            ),
+        pyarrow.array(
+            [dt.date(2021, 9, 27), None, dt.date(2011, 9, 27)], type=pyarrow.date32(),
         ),
     ),
-)
+    (
+        pandas.Series(
+            [dt.date(1677, 9, 22), dt.date(1970, 1, 1), dt.date(2262, 4, 11)],
+            dtype="dbdate",
+        ),
+        pyarrow.array(
+            [dt.date(1677, 9, 22), dt.date(1970, 1, 1), dt.date(2262, 4, 11)],
+            type=pyarrow.date32(),
+        ),
+    ),
+    (pandas.Series([], dtype="dbtime"), pyarrow.array([], type=pyarrow.time64("ns")),),
+    (
+        pandas.Series([None, None, None], dtype="dbtime"),
+        pyarrow.array([None, None, None], type=pyarrow.time64("ns")),
+    ),
+    (
+        pandas.Series(
+            [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_999)], dtype="dbtime",
+        ),
+        pyarrow.array(
+            [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_999)],
+            type=pyarrow.time64("ns"),
+        ),
+    ),
+    (
+        pandas.Series(
+            [
+                dt.time(0, 0, 0, 0),
+                dt.time(12, 30, 15, 125_000),
+                dt.time(23, 59, 59, 999_999),
+            ],
+            dtype="dbtime",
+        ),
+        pyarrow.array(
+            [
+                dt.time(0, 0, 0, 0),
+                dt.time(12, 30, 15, 125_000),
+                dt.time(23, 59, 59, 999_999),
+            ],
+            type=pyarrow.time64("ns"),
+        ),
+    ),
+]
+SERIES_ARRAYS_CUSTOM_ARROW_TYPES = [
+    (pandas.Series([], dtype="dbdate"), pyarrow.array([], type=pyarrow.date64())),
+    (
+        pandas.Series([None, None, None], dtype="dbdate"),
+        pyarrow.array([None, None, None], type=pyarrow.date64()),
+    ),
+    (
+        pandas.Series(
+            [dt.date(2021, 9, 27), None, dt.date(2011, 9, 27)], dtype="dbdate"
+        ),
+        pyarrow.array(
+            [dt.date(2021, 9, 27), None, dt.date(2011, 9, 27)], type=pyarrow.date64(),
+        ),
+    ),
+    (
+        pandas.Series(
+            [dt.date(1677, 9, 22), dt.date(1970, 1, 1), dt.date(2262, 4, 11)],
+            dtype="dbdate",
+        ),
+        pyarrow.array(
+            [dt.date(1677, 9, 22), dt.date(1970, 1, 1), dt.date(2262, 4, 11)],
+            type=pyarrow.date64(),
+        ),
+    ),
+    (pandas.Series([], dtype="dbtime"), pyarrow.array([], type=pyarrow.time32("ms")),),
+    (
+        pandas.Series([None, None, None], dtype="dbtime"),
+        pyarrow.array([None, None, None], type=pyarrow.time32("ms")),
+    ),
+    (
+        pandas.Series(
+            [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_000)], dtype="dbtime",
+        ),
+        pyarrow.array(
+            [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_000)],
+            type=pyarrow.time32("ms"),
+        ),
+    ),
+    (
+        pandas.Series(
+            [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_999)], dtype="dbtime",
+        ),
+        pyarrow.array(
+            [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_999)],
+            type=pyarrow.time64("us"),
+        ),
+    ),
+    (
+        pandas.Series(
+            [
+                dt.time(0, 0, 0, 0),
+                dt.time(12, 30, 15, 125_000),
+                dt.time(23, 59, 59, 999_999),
+            ],
+            dtype="dbtime",
+        ),
+        pyarrow.array(
+            [
+                dt.time(0, 0, 0, 0),
+                dt.time(12, 30, 15, 125_000),
+                dt.time(23, 59, 59, 999_999),
+            ],
+            type=pyarrow.time64("us"),
+        ),
+    ),
+]
+
+
+@pytest.mark.parametrize(("series", "expected"), SERIES_ARRAYS_DEFAULT_TYPES)
 def test_to_arrow(series, expected):
     array = pyarrow.array(series)
     assert array.equals(expected)
 
 
-@pytest.mark.parametrize(
-    ("series", "expected"),
-    (
-        (pandas.Series([], dtype="dbdate"), pyarrow.array([], type=pyarrow.date64())),
-        (
-            pandas.Series([None, None, None], dtype="dbdate"),
-            pyarrow.array([None, None, None], type=pyarrow.date64()),
-        ),
-        (
-            pandas.Series(
-                [dt.date(2021, 9, 27), None, dt.date(2011, 9, 27)], dtype="dbdate"
-            ),
-            pyarrow.array(
-                [dt.date(2021, 9, 27), None, dt.date(2011, 9, 27)],
-                type=pyarrow.date64(),
-            ),
-        ),
-        (
-            pandas.Series(
-                [dt.date(1677, 9, 22), dt.date(1970, 1, 1), dt.date(2262, 4, 11)],
-                dtype="dbdate",
-            ),
-            pyarrow.array(
-                [dt.date(1677, 9, 22), dt.date(1970, 1, 1), dt.date(2262, 4, 11)],
-                type=pyarrow.date64(),
-            ),
-        ),
-        (
-            pandas.Series([], dtype="dbtime"),
-            pyarrow.array([], type=pyarrow.time32("ms")),
-        ),
-        (
-            pandas.Series([None, None, None], dtype="dbtime"),
-            pyarrow.array([None, None, None], type=pyarrow.time32("ms")),
-        ),
-        (
-            pandas.Series(
-                [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_000)],
-                dtype="dbtime",
-            ),
-            pyarrow.array(
-                [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_000)],
-                type=pyarrow.time32("ms"),
-            ),
-        ),
-        (
-            pandas.Series(
-                [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_999)],
-                dtype="dbtime",
-            ),
-            pyarrow.array(
-                [dt.time(0, 0, 0, 0), None, dt.time(23, 59, 59, 999_999)],
-                type=pyarrow.time64("us"),
-            ),
-        ),
-        (
-            pandas.Series(
-                [
-                    dt.time(0, 0, 0, 0),
-                    dt.time(12, 30, 15, 125_000),
-                    dt.time(23, 59, 59, 999_999),
-                ],
-                dtype="dbtime",
-            ),
-            pyarrow.array(
-                [
-                    dt.time(0, 0, 0, 0),
-                    dt.time(12, 30, 15, 125_000),
-                    dt.time(23, 59, 59, 999_999),
-                ],
-                type=pyarrow.time64("us"),
-            ),
-        ),
-    ),
-)
+@pytest.mark.parametrize(("series", "expected"), SERIES_ARRAYS_CUSTOM_ARROW_TYPES)
 def test_to_arrow_w_arrow_type(series, expected):
     array = pyarrow.array(series, type=expected.type)
     assert array.equals(expected)
+
+
+@pytest.mark.parametrize(
+    ["expected", "pyarrow_array"],
+    SERIES_ARRAYS_DEFAULT_TYPES + SERIES_ARRAYS_CUSTOM_ARROW_TYPES,
+)
+def test_from_arrow(pyarrow_array: pyarrow.Array, expected: pandas.Series):
+    # Convert to RecordBatch because types_mapper argument is ignored when
+    # using a pyarrow.Array. https://issues.apache.org/jira/browse/ARROW-9664
+    record_batch = pyarrow.RecordBatch.from_arrays([pyarrow_array], ["test_col"])
+    dataframe = record_batch.to_pandas(date_as_object=False, types_mapper=types_mapper)
+    series = dataframe["test_col"]
+    pandas.testing.assert_extension_array_equal(series, expected)

--- a/tests/unit/test_arrow.py
+++ b/tests/unit/test_arrow.py
@@ -179,10 +179,68 @@ def test_to_arrow_w_arrow_type(series, expected):
     ["expected", "pyarrow_array"],
     SERIES_ARRAYS_DEFAULT_TYPES + SERIES_ARRAYS_CUSTOM_ARROW_TYPES,
 )
-def test_from_arrow(pyarrow_array: pyarrow.Array, expected: pandas.Series):
+def test_series_from_arrow(pyarrow_array: pyarrow.Array, expected: pandas.Series):
     # Convert to RecordBatch because types_mapper argument is ignored when
     # using a pyarrow.Array. https://issues.apache.org/jira/browse/ARROW-9664
     record_batch = pyarrow.RecordBatch.from_arrays([pyarrow_array], ["test_col"])
     dataframe = record_batch.to_pandas(date_as_object=False, types_mapper=types_mapper)
     series = dataframe["test_col"]
     pandas.testing.assert_series_equal(series, expected, check_names=False)
+
+
+@pytest.mark.parametrize(
+    ["expected", "pyarrow_array"],
+    SERIES_ARRAYS_DEFAULT_TYPES + SERIES_ARRAYS_CUSTOM_ARROW_TYPES,
+)
+def test_series_from_arrow_scalars(
+    pyarrow_array: pyarrow.Array, expected: pandas.Series
+):
+    scalars = []
+    for scalar in pyarrow_array:
+        scalars.append(scalar)
+        assert isinstance(scalar, pyarrow.Scalar)
+    series = pandas.Series(scalars, dtype=expected.dtype)
+    pandas.testing.assert_series_equal(series, expected)
+
+
+def test_dataframe_from_arrow():
+    record_batch = pyarrow.RecordBatch.from_arrays(
+        [
+            pyarrow.array(
+                [dt.date(2021, 11, 4), dt.date(2038, 1, 20), None, dt.date(1970, 1, 1)],
+                type=pyarrow.date32(),
+            ),
+            pyarrow.array(
+                [
+                    dt.time(10, 7, 8, 995_325),
+                    dt.time(23, 59, 59, 999_999),
+                    None,
+                    dt.time(0, 0, 0, 0),
+                ],
+                type=pyarrow.time64("us"),
+            ),
+            pyarrow.array([1, 2, 3, 4]),
+        ],
+        ["date_col", "time_col", "int_col"],
+    )
+    dataframe = record_batch.to_pandas(date_as_object=False, types_mapper=types_mapper)
+    expected = pandas.DataFrame(
+        {
+            "date_col": pandas.Series(
+                [dt.date(2021, 11, 4), dt.date(2038, 1, 20), None, dt.date(1970, 1, 1)],
+                dtype="dbdate",
+            ),
+            "time_col": pandas.Series(
+                [
+                    dt.time(10, 7, 8, 995_325),
+                    dt.time(23, 59, 59, 999_999),
+                    None,
+                    dt.time(0, 0, 0, 0),
+                ],
+                dtype="dbtime",
+            ),
+            "int_col": [1, 2, 3, 4],
+        },
+        columns=["date_col", "time_col", "int_col"],
+    )
+    pandas.testing.assert_frame_equal(dataframe, expected)

